backend/cfg/vectorize.ml

```diff
@@ -134,7 +138,327 @@ module Dependency_graph = struct
+      let addressing_mode_2 = t2.addressing_mode in
+      let args_1 = args t1 in
+      let args_2 = args t2 in
+      match addressing_mode_1, addressing_mode_2 with
+      | Ibased (symbol1, global1, n1), Ibased (symbol2, global2, n2) -> (
+        match global1, global2 with
+        | Global, Global | Local, Local ->
+          if symbol1 = symbol2 then Some (n2 - n1) else None
+        | Global, Local | Local, Global -> None)
+      | Iindexed n1, Iindexed n2 ->
+        let arg0_1 = Array.get args_1 0 in
+        let arg0_2 = Array.get args_2 0 in
+        let arg0_compare = arg0_1.stamp - arg0_2.stamp in
+        if arg0_compare = 0 then Some (n2 - n1) else None
+      | Iindexed2 n1, Iindexed2 n2 ->
+        let arg0_1 = Array.get args_1 0 in
+        let arg0_2 = Array.get args_2 0 in
+        let arg0_compare = arg0_1.stamp - arg0_2.stamp in
+        if arg0_compare = 0
+        then
+          let arg1_1 = Array.get args_1 1 in
+          let arg1_2 = Array.get args_2 1 in
+          let arg1_compare = arg1_1.stamp - arg1_2.stamp in
+          if arg1_compare = 0 then Some (n2 - n1) else None
+        else None
+      | Iscaled (scale1, n1), Iscaled (scale2, n2) ->
+        let arg0_1 = Array.get args_1 0 in
+        let arg0_2 = Array.get args_2 0 in
+        let arg0_compare = arg0_1.stamp - arg0_2.stamp in
+        if arg0_compare = 0
+        then
+          let scale_compare = scale1 - scale2 in
+          if scale_compare = 0 then Some (n2 - n1) else None
+        else None
+      | Iindexed2scaled (scale1, n1), Iindexed2scaled (scale2, n2) ->
+        let arg0_1 = Array.get args_1 0 in
+        let arg0_2 = Array.get args_2 0 in
+        let arg0_compare = arg0_1.stamp - arg0_2.stamp in
+        if arg0_compare = 0
+        then
+          let arg1_1 = Array.get args_1 1 in
+          let arg1_2 = Array.get args_2 1 in
+          let arg1_compare = arg1_1.stamp - arg1_2.stamp in
+          if arg1_compare = 0
+          then
+            let scale_compare = scale1 - scale2 in
+            if scale_compare = 0 then Some (n2 - n1) else None
+          else None
+        else None
+      | Ibased _, _ -> None
+      | Iindexed _, _ -> None
+      | Iindexed2 _, _ -> None
+      | Iscaled _, _ -> None
+      | Iindexed2scaled _, _ -> None
+
+    let width_of (memory_chunk : Cmm.memory_chunk) =
+      match memory_chunk with
+      | Byte_unsigned | Byte_signed -> 1
+      | Sixteen_unsigned | Sixteen_signed -> 2
+      | Thirtytwo_unsigned | Thirtytwo_signed | Single _ -> 4
+      | Word_int | Word_val | Double -> 8
+      | Onetwentyeight_unaligned | Onetwentyeight_aligned -> 16
+
+    let is_adjacent (t1 : t) (t2 : t) =
+      let res =
+        if Cmm.equal_memory_chunk t1.memory_chunk t2.memory_chunk
+        then
+          let width = width_of t1.memory_chunk in
+          let offset_option = offset_of t1 t2 in
+          match offset_option with
+          | None -> false
+          | Some offset -> width = offset
```

Comment:
does it detect adjacent loads where the first load access a higher address and then a lower address: 
```
x <- load a + 8
y <- load a
```

Reply:
yes because I sort the instructions before doing this